docs/design/2020-06-27-non-recursive-common-table-expression.md

```diff
@@ -0,0 +1,143 @@
+  cte2 AS (SELECT c, d FROM table2)
+SELECT /*+ MERGE(cte1) NO_MERGE(cte2) */ cte1.b, cte2.d
+FROM cte1 JOIN cte2
+WHERE cte1.a = cte2.c;
+```
+_(cte1 should be merged and cte2 should be materialized)_
+
+
+
+## Rationale
+
+Mainstream database systems utilize `Merge` and `Materialization` to implement CTE.
+
+In these DBMS, `Merge` is preferred when the CTE has no side-effect, and is referenced in a very few times. `Materialization` should be avoid if possible.
+
+If `Materialization` is necessary, these optimizations can be applied:
+1. Materialized in memory [13].
+2. Postpones until its contents are needed [1].
+3. Add index(es) according to access method [1].
+4. Pushdown predicates, and pushdown disjunction of all predicates if referenced multiple times [4][5].
+
+* MySQL using these two strategies: 1) Merge the derived table into the outer query block, or 2) Materialize the derived table to an internal temporary table [1]. The optimizer avoids unnecessary materialization whenever possible, which enables pushing down conditions from the outer query to derived tables and produces more efficient execution plans [1].
+
+* MariaDB using `Materialization` as basic algorithm. To optimize performance, MariaDB using `Merge` strategy, and utilizes "Condition Pushdown" [5].
+
+* Greenplum also using `Merge` and `Materialization`, and utilize rules and cost-bases algorithm to optimize execution plan. Moreover, Greenplum introduces a producer-consumer model to implement `Materialization` [4].
+
+* Postgresql have always used `Materialization` strategy, and introduced `Merge` in a recent commit [9][10].
+
+* CockroachDB uses `Materialization` strategy by default, but chooses `Merge` strategy when 1) see `NO MATERIALIZED` hint, or 2) the CTE has no side-effect and is referenced only once [11][12].
+
+
+
+## Implementation
+
+1. Stage 1: `Merge` should be not difficult. A rewrite rule is enough, similar to rewriting subquery [2].
+2. Stage 2: Implement `Materialization`, by storing chunks temporarily in memory or disk.
+3. Stage 3: `Predicate Push-down` for `Materialization`, and estimate cost for both strategies.
+
+
+## Open issues
+[TiDB#17472: support common table expression](https://github.com/pingcap/tidb/issues/17472)
+
+## References
+1. [MySQL 8.0 Reference Manual, 8.2.2.4 Optimizing Derived Tables, View References, and Common Table Expressions with Merging or Materialization](https://dev.mysql.com/doc/refman/8.0/en/derived-table-optimization.html)
+2. [Subquery Optimization in TiDB](https://pingcap.com/blog/2016-12-07-Subquery-Optimization-in-TiDB/)
+3. [Consider using a disk-based hash table for hash join avoiding OOM](https://github.com/pingcap/tidb/issues/11607)
+4. [Optimization of Common Table Expressions in MPP Database Systems](http://www.vldb.org/pvldb/vol8/p1704-elhelw.pdf)
+5. [Implementing Common Table Expressions for MariaDB](https://seim-conf.org/media/materials/2017/proceedings/SEIM-2017_Full_Papers.pdf#page=13)
```

Comment:
how about this link: http://ceur-ws.org/Vol-1864/paper_6.pdf, it's only 6 pages that contain only the required content.